src/sources/kubernetes/mod.rs

```diff
@@ -300,6 +267,93 @@ fn transform_file() -> crate::Result<Box<dyn Transform>> {
     })
 }
 
+/// Contains several regexes that can parse common forms of pod_uid.
+/// On first message, regexes are tryed out one after the other until
+/// first succesfull one has been found. After that that regex will be
+/// always used.
+///
+/// If nothing succeds the message is still passed.
+fn transform_pod_uid() -> crate::Result<ApplicableTransform> {
+    let mut regexes = Vec::new();
+
+    let namespace_regex = r"(?P<pod_namespace>[0-9a-z.\-]*)";
+    let name_regex = r"(?P<pod_name>[0-9a-z.\-]*)";
+    let uid_regex = r"(?P<object_uid>([0-9A-Fa-f]{8}[-][0-9A-Fa-f]{4}[-][0-9A-Fa-f]{4}[-][0-9A-Fa-f]{4}[-][0-9A-Fa-f]{12}|[0-9A-Fa-f]{32}))";
+
+    // Definition of pod_uid has been well defined since Kubernetes 1.14 with https://github.com/kubernetes/kubernetes/pull/74441
+
+    // Minikube 1.15, MicroK8s 1.15,1.14,1.16 , DigitalOcean 1.16 , Google Kubernetes Engine 1.13, 1.14, EKS 1.14
+    // format: namespace_name_UID
+    regexes.push(format!(
+        "^{}_{}_{}$",
+        namespace_regex, name_regex, uid_regex
+    ));
+
+    // EKS 1.13 , AKS 1.13.12, MicroK8s 1.13
+    // If everything else fails, try to at least parse out uid from somewhere.
+    // This is somewhat robust as UUID format is hard to create by accident
+    // ,at least in this context, plus regex requires that UUID is separated
+    // from other data either by start,end of string or by non UUID character.
+    regexes.push(format!(
+        r"(^|[^0-9A-Fa-f\-]){}([^0-9A-Fa-f\-]|$)",
+        uid_regex
+    ));
+
+    let mut transforms = Vec::new();
+    for regex in regexes {
+        let mut config = RegexParserConfig::default();
+
+        config.field = Some("pod_uid".into());
+        config.regex = regex;
+        // Remove pod_uid as it isn't usable anywhere else.
+        config.drop_field = true;
+        config.drop_failed = true;
+
+        let transform = RegexParser::build(&config).map_err(|e| {
+            format!(
+                "Failed in creating pod_uid regex transform with error: {:?}",
+                e
+            )
+        })?;
+        transforms.push(transform);
+    }
+
+    Ok(ApplicableTransform::Candidates(transforms))
+}
+
+/// Contains several transforms. On first message, transforms are tryed
+/// out one after the other until first succesfull one has been found.
+/// After that that transform will always be used.
+///
+/// If nothing succeds the message is still passed.
+enum ApplicableTransform {
```

Comment:
I am trying to understand the use of this, what is the benefit of this over optimistically trying all the transforms all the time? They can but don't need to modify the event.

Reply:
While that's true, this is more efficient as there is only a cost of one switch statement after first event. While in the case of optimistically trying all the transforms, if it succeeds on the first transform, all is good. But if not, processing of all events will be more expensive as inapplicable transforms would have to do some computation to establish that they can't do it. 

This situation/pattern occurs twice in the code.